lighthouse-core/lib/i18n/i18n.js

```diff
@@ -358,143 +344,140 @@ function getRendererFormattedStrings(locale) {
+   */
+  const getIcuMessageFn = (message, values) => {
+    const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === message);
+    if (!keyname) throw new Error(`Could not locate: ${message}`);
 
     const filenameToLookup = keyname in fileStrings ? filename : __filename;
     const unixStyleFilename = path.relative(LH_ROOT, filenameToLookup).replace(/\\/g, '/');
-    const icuMessageId = `${unixStyleFilename} | ${keyname}`;
-    const icuMessageInstances = _icuMessageInstanceMap.get(icuMessageId) || [];
+    const i18nId = `${unixStyleFilename} | ${keyname}`;
 
-    let indexOfInstance = icuMessageInstances.findIndex(inst => isDeepEqual(inst.values, values));
-    if (indexOfInstance === -1) {
-      icuMessageInstances.push({icuMessageId, icuMessage, values});
-      indexOfInstance = icuMessageInstances.length - 1;
-    }
-
-    _icuMessageInstanceMap.set(icuMessageId, icuMessageInstances);
-
-    return `${icuMessageId} # ${indexOfInstance}`;
+    return {
+      i18nId,
+      values,
+      formattedDefault: _formatMessage(message, values, DEFAULT_LOCALE),
+    };
   };
 
-  return getMessageInstanceIdFn;
+  return getIcuMessageFn;
 }
 
 /**
- * Returns true if string is an ICUMessage reference.
- * @param {string} icuMessageIdOrRawString
- * @return {boolean}
+ * Returns whether `icuMessageOrNot`` is an `LH.IcuMessage` instance.
+ * @param {unknown} icuMessageOrNot
+ * @return {icuMessageOrNot is LH.IcuMessage}
  */
-function isIcuMessage(icuMessageIdOrRawString) {
-  return MESSAGE_INSTANCE_ID_QUICK_REGEX.test(icuMessageIdOrRawString) &&
-      MESSAGE_INSTANCE_ID_REGEX.test(icuMessageIdOrRawString);
-}
+function isIcuMessage(icuMessageOrNot) {
+  if (!isObjectOfUnknownValues(icuMessageOrNot)) {
+    return false;
+  }
 
-/**
- * @param {string} icuMessageIdOrRawString
- * @param {LH.Locale} locale
- * @return {string}
- */
-function getFormatted(icuMessageIdOrRawString, locale) {
-  if (isIcuMessage(icuMessageIdOrRawString)) {
-    return _resolveIcuMessageInstanceId(icuMessageIdOrRawString, locale).formattedString;
+  const {i18nId, values, formattedDefault} = icuMessageOrNot;
+  if (typeof i18nId !== 'string') {
+    return false;
   }
 
-  return icuMessageIdOrRawString;
-}
+  // formattedDefault is required.
+  if (typeof formattedDefault !== 'string') {
+    return false;
+  }
 
-/**
- * @param {LH.Locale} locale
- * @param {string} icuMessageId
- * @param {Record<string, string | number>} [values]
- * @return {string}
- */
-function getFormattedFromIdAndValues(locale, icuMessageId, values) {
-  const icuMessageIdRegex = /(.* \| .*)$/;
-  if (!icuMessageIdRegex.test(icuMessageId)) throw new Error('This is not an ICU message ID');
+  // Values is optional.
+  if (values !== undefined) {
+    if (!isObjectOfUnknownValues(values)) {
+      return false;
+    }
+    for (const value of Object.values(values)) {
+      if (typeof value !== 'string' && typeof value !== 'number') {
+        return false;
```

Comment:
what are the consequences if we're accidentally passing an undefined or null or something in here? presumably it should be impossible due to types, but what if :)

Reply:
> what are the consequences if we're accidentally passing an undefined or null or something in here? presumably it should be impossible due to types, but what if :)

it'll end up throwing in `_preformatValues` when it tries to insert the null/undefined into the message but it only allows strings (inserted unchanged) or numbers (inserted formatted).

We could do a check in `createMessageInstanceIdFn`/`createIcuMessageFn` so it'll throw closer to the root cause (immediately on calling `_str()`) but throwing in `_preformatValues` is what happens in the current LH too.

Reply:
got it, yeah no need for changes here, just wondering if this introduced anything new to worry about 👍